src/size_hint.rs

```diff
@@ -1,149 +1,493 @@
+    fn from(range: ops::RangeFrom<usize>) -> Self {
+        Self {
+            lower: range.start,
+            upper: Upper::Unbounded,
+            out_of_fuel: false,
+        }
+    }
+}
+impl From<ops::RangeInclusive<usize>> for SizeHint {
+    fn from(range: ops::RangeInclusive<usize>) -> Self {
+        Self {
+            lower: *range.start(),
+            upper: Upper::Bounded(*range.end()),
+            out_of_fuel: false,
+        }
+    }
+}
+
+impl ops::RangeBounds<usize> for SizeHint {
+    fn start_bound(&self) -> ops::Bound<&usize> {
+        ops::Bound::Included(&self.lower)
+    }
+
+    fn end_bound(&self) -> ops::Bound<&usize> {
+        match &self.upper {
+            Upper::Bounded(upper) => ops::Bound::Included(upper),
+            Upper::Unbounded | Upper::Unknown => ops::Bound::Unbounded,
+        }
+    }
+}
+
+impl ops::Add for SizeHint {
+    type Output = Self;
+    /// Identical to [`SizeHint::and()`].
+    fn add(self, rhs: Self) -> Self::Output {
+        self.and(rhs)
+    }
+}
+impl ops::AddAssign for SizeHint {
+    /// Identical to [`SizeHint::and()`].
+    fn add_assign(&mut self, rhs: Self) {
+        *self = self.and(rhs)
+    }
+}
+impl ops::BitOr for SizeHint {
+    type Output = Self;
+    /// Identical to [`SizeHint::or()`].
+    fn bitor(self, rhs: Self) -> Self::Output {
+        self.or(rhs)
+    }
 }
 
-/// Take the maximum of the `lhs` and `rhs` size hints.
+/// Context passed recursively through [`Arbitrary::size_hint()`] implementations.
 ///
-/// If `hints` is empty, returns `(0, Some(0))`, aka the size of consuming
-/// nothing.
-#[inline]
-pub fn or_all(hints: &[(usize, Option<usize>)]) -> (usize, Option<usize>) {
-    if let Some(head) = hints.first().copied() {
-        hints[1..].iter().copied().fold(head, or)
-    } else {
-        (0, Some(0))
+/// Used to ensure that `size_hint()` cannot execute unbounded recursion through generic types,
+/// by terminating the recursion after a certain number of operations, measured by the “fuel” value
+/// in this context.
+///
+/// When you are implementing `size_hint()` for a generic type, use this context by calling
+/// [`Context::get()`] instead of calling other `size_hint()` implementations directly,
+/// and use `and()`, `or()`, `and_all()`, and `or_all()` to compose size hints.
+pub struct Context {
+    fuel: Cell<u32>,
+}
+
+impl Default for Context {
+    fn default() -> Self {
+        Self::new(100)
     }
 }
 
+impl Context {
+    /// Creates a context which allows no more than `fuel` recursive calls.
+    pub fn new(fuel: u32) -> Self {
+        Context {
+            fuel: Cell::new(fuel),
+        }
+    }
+
+    /// Returns [`T::size_hint()`][Arbitrary::size_hint], and consumes 1 unit of fuel too.
+    #[inline]
+    pub fn get<'a, T: Arbitrary<'a>>(&self) -> SizeHint {
+        if self.consume_fuel() {
+            T::size_hint(self)
+        } else {
+            SizeHint::OUT_OF_FUEL
+        }
+    }
```

Comment:
Two nitpicks:

1. I don't think we should make guarantees about how much fuel any particular method consumes. When describing what fuel is (say on the `Context::new` constructor), we can give rough estimates, but I think we want to keep it somewhat fuzzy so that we can adjust if necessary. The important part is that it provides a finite bound to size-hint computation, not *exactly* what that bound is.
2. As discussed in thread, we shouldn't actually need this method anymore, once `SizeHint` construction (even for leaf types) consumes fuel. Users can just call the trait methods directly instead.